### DISCUSSION TITLE
Not able to flash crazyflie drone.

### DISCUSSION THREAD

**Discussion by cuber-1** [bitcraze/discussions#1460] [General]
I'm currently trying to run the crazyflie drone through my Windows computer. I'm not using a VM and am trying to install the firmware to my computer. I am using this link: https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/building-and-flashing/build/. 
I am in the compiling section currently and have ran 

$ make cf2_defconfig
$ make -j 12

The build was successful and showed CLEAN in the command prompt. 

I am in this section now: 

------
Automatically enter bootloader mode
Make sure the Crazyflie is on
In your terminal, run CLOAD_CMDS="-w [CRAZYFLIE_URI]" make cload
or run cfloader flash build/cf2.bin stm32-fw -w [CRAZYFLIE_URI] with [CRAZYFLIE_URI] being the uri of the crazyflie.
------

I've tried to run both the CLOUD_CMDS command and the cfloader flash command but neither works. 

For CLOUD_CMDS:
root@NAME:/mnt/c/Users/NAME/crazyflie-firmware# CLOAD_CMDS="-w radio://0/80/2M" make cload
tools/kbuild/Makefile.kbuild:147: warning: overriding recipe for target 'cload'
Makefile:140: warning: ignoring old recipe for target 'cload'
make[1]: Entering directory '/mnt/c/Users/NAME/crazyflie-firmware/build'
python.exe -m cfloader -w radio://0/80/2M flash  cf2.bin stm32-fw
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\NAME\AppData\Local\Programs\Python\Python312\Lib\site-packages\cfloader\__main__.py", line 1, in <module>
    from cfloader import main
ImportError: cannot import name 'main' from 'cfloader' (C:\Users\NAME\AppData\Local\Programs\Python\Python312\Lib\site-packages\cfloader\__init__.py)
make[1]: *** [/mnt/c/Users/NAME/crazyflie-firmware/Makefile:140: cload] Error 1
make[1]: Leaving directory '/mnt/c/Users/NAME/crazyflie-firmware/build'
make: *** [tools/kbuild/Makefile.kbuild:150: sub-make] Error 2

For cfloader:
root@NAME:/mnt/c/Users/NAME/crazyflie-firmware# cfloader flash build/cf2.bin stm32-fw -w radio://0/80/2M
cfloader: command not found
root@NAME:/mnt/c/Users/NAME/crazyflie-firmware#

For the cfloader, however, when I do "pip show cfloader", it shows: 

root@NAME:/mnt/c/Users/NAME/crazyflie-firmware# pip show cfloader
Name: cfloader
Version: 0.1.4
Summary: Loads python modules from a configuration file.
Home-page: https://github.com/shachibista/cfloader
Author: Shachi Bista
Author-email: shachibista@gmail.com
License: MIT license
Location: /usr/local/lib/python3.10/dist-packages
Requires:
Required-by:
root@NAME:/mnt/c/Users/NAME/crazyflie-firmware#

What do I do?

**Comment by ataffanel** [comment]
Under WSL, the makefile tries to run the Windows python ```python.exe -m cfloader -w radio://0/80/2M flash cf2.bin stm32-fw```. So you must make sure that cfclient is installed on the windows side by running ```pip install cfclient``` in a Windows shell or with ```pip.exe```. This is the relevant part of the documentation: https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/building-and-flashing/build/#windows

**Reply by cuber-1** [reply]
I believe my cfclient is already installed on the windows side. I just type cfclient into no specified directory and the application pops open and works great. I've been able to connect and fly the drone before. Just to be safe I tried again: 

```
C:\Users\NAME>pip install cfclient
Requirement already satisfied: cfclient in c:\users\NAME\appdata\local\programs\python\python312\lib\site-packages (2024.7.1)
Requirement already satisfied: cflib>=0.1.26 in c:\users\NAME\appdata\local\programs\python\python312\lib\site-packages (from cfclient) (0.1.26)
Requirement already satisfied: setuptools in c:\users\NAME\appdata\local\programs\python\python312\lib\site-packages (from cfclient) (70.2.0)
Requirement already satisfied: appdirs~=1.4.0 in c:\users\NAME\appdata\local\programs\python\python312\lib\site-packages (from cfclient) (1.4.4)
Requirement already satisfied: pyzmq~=25.0 in c:\users\NAME\appdata\local\programs\python\python312\lib\site-packages (from cfclient) (25.1.2)
Requirement already satisfied: pyqtgraph~=0.13 in c:\users\NAME\appdata\local\programs\python\python312\lib\site-packages (from cfclient) (0.13.7)
Requirement already satisfied: PyYAML~=6.0.1 in c:\users\NAME\appdata\local\programs\python\python312\lib\site-packages (from cfclient) (6.0.1)
Requirement already satisfied: numpy~=1.20 in c:\users\NAME\appdata\local\programs\python\python312\lib\site-packages (from cfclient) (1.26.4)
Requirement already satisfied: vispy~=0.14.2 in c:\users\NAME\appdata\local\programs\python\python312\lib\site-packages (from cfclient) (0.14.3)
Requirement already satisfied: pyopengl~=3.1.7 in c:\users\NAME\appdata\local\programs\python\python312\lib\site-packages (from cfclient) (3.1.7)
Requirement already satisfied: pyserial~=3.5 in c:\users\NAME\appdata\local\programs\python\python312\lib\site-packages (from cfclient) (3.5)
Requirement already satisfied: pyqt6>=6.4 in c:\users\NAME\appdata\local\programs\python\python312\lib\site-packages (from cfclient) (6.7.0)
Requirement already satisfied: PyQt6-sip>=13.4 in c:\users\NAME\appdata\local\programs\python\python312\lib\site-packages (from cfclient) (13.6.0)
Requirement already satisfied: pysdl2~=0.9.14 in c:\users\NAME\appdata\local\programs\python\python312\lib\site-packages (from cfclient) (0.9.16)
Requirement already satisfied: pysdl2-dll==2.24.0 in c:\users\NAME\appdata\local\programs\python\python312\lib\site-packages (from cfclient) (2.24.0)
Requirement already satisfied: pyusb>=1.0.0b2 in c:\users\NAME\appdata\local\programs\python\python312\lib\site-packages (from cflib>=0.1.26->cfclient) (1.2.1)
Requirement already satisfied: libusb-package~=1.0 in c:\users\NAME\appdata\local\programs\python\python312\lib\site-packages (from cflib>=0.1.26->cfclient) (1.0.26.1)
Requirement already satisfied: scipy~=1.7 in c:\users\NAME\appdata\local\programs\python\python312\lib\site-packages (from cflib>=0.1.26->cfclient) (1.14.0)
Requirement already satisfied: PyQt6-Qt6<6.8.0,>=6.7.0 in c:\users\NAME\appdata\local\programs\python\python312\lib\site-packages (from pyqt6>=6.4->cfclient) (6.7.2)
Requirement already satisfied: freetype-py in c:\users\NAME\appdata\local\programs\python\python312\lib\site-packages (from vispy~=0.14.2->cfclient) (2.4.0)
Requirement already satisfied: hsluv in c:\users\NAME\appdata\local\programs\python\python312\lib\site-packages (from vispy~=0.14.2->cfclient) (5.0.4)
Requirement already satisfied: kiwisolver in c:\users\NAME\appdata\local\programs\python\python312\lib\site-packages (from vispy~=0.14.2->cfclient) (1.4.5)
Requirement already satisfied: packaging in c:\users\NAME\appdata\local\programs\python\python312\lib\site-packages (from vispy~=0.14.2->cfclient) (24.1)
```

Now that you bring this section up, however, I do remember doing something that could have an impact, but I'm not sure. In the Windows instructions, one part asks to follow the instructions made for Ubuntu. After I completed the Windows and Ubuntu instructions, I began to follow the instructions in compile section. 

I ran the following:

```
root@NAME:/mnt/c/Users/NAME/crazyflie-firmware# make cf2_defconfig
make[1]: Entering directory '/mnt/c/Users/NAME/crazyflie-firmware/build'
  HOSTCC  scripts/basic/fixdep
/bin/sh: 1: gcc: not found
make[2]: *** [scripts/Makefile.host:91: scripts/basic/fixdep] Error 127
make[1]: *** [tools/kbuild/Makefile.kbuild:318: scripts_basic] Error 2
make[1]: Leaving directory '/mnt/c/Users/name/crazyflie-firmware/build'
make: *** [tools/kbuild/Makefile.kbuild:150: sub-make] Error 2
```

It said the gcc was not found, and ChatGPT said to run this 
```
sudo apt install gcc
```
Everything worked fine after that, and the build showed CLEAN

Does this have anything to do with the issue?

**Comment by cuber-1** [comment]
Here is an update: I deleted all files and installed everything again. I tried flashing the sample code found here: https://www.bitcraze.io/documentation/tutorials/getting-started-with-development/. Here is what happens when I try to run the build: 

root@NAME:/mnt/c/Users/NAME/crazyflie-firmware# make -j 12
make[1]: Entering directory '/mnt/c/Users/NAME/crazyflie-firmware/build'
  GEN     ./Makefile
scripts/kconfig/conf  --silentoldconfig Kconfig
  Using .. as source for kernel
  GEN     ./Makefile
  CC      src/init/main.o
  CC      src/utils/src/cfassert.o
  CC      src/hal/src/freeRTOSdebug.o
  CC      src/hal/src/ledseq.o
  CC      src/hal/src/pm_stm32f4.o
python3 ../tools/make/versionTemplate.py --crazyflie-base .. ../src/utils/src/version.vtpl src/utils/src/version_gen.c
  CC      src/hal/src/radiolink.o
  CC      src/hal/src/sensors_bmi088_bmp3xx.o
  CC      src/drivers/src/led.o
  LD      src/init/built-in.o
  CC      src/deck/drivers/src/usddeck.o
  CC      src/hal/src/sensors_mpu9250_lps25h.o
  CC      src/drivers/src/nvic.o
  CC      src/modules/src/serial_4way.o
  CC      src/hal/src/usblink.o
  CC      src/modules/src/system.o
  CC      src/modules/src/cpx/cpxlink.o
  LD      src/drivers/src/built-in.o
  LD      src/deck/drivers/src/built-in.o
  LD      src/hal/src/built-in.o
  LD      src/drivers/built-in.o
  LD      src/hal/built-in.o
  LD      src/deck/built-in.o
  LD      src/modules/src/cpx/built-in.o
  LD      src/modules/src/built-in.o
  LD      src/modules/built-in.o
  CC      src/utils/src/version_gen.o
  LD      src/utils/src/version.o
  LD      src/utils/src/built-in.o
  LD      src/utils/built-in.o
  LD      src/built-in.o
  LD      firmware
  COPY  cf2.bin
  COPY  cf2.hex
Build for the cf2 platform!
Build 161:ec6a7929d5b1 (2023.11 +161) root@NAME:/mnt/c/Users/NAME/crazyflie-firmware# make -j 12
make[1]: Entering directory '/mnt/c/Users/NAME/crazyflie-firmware/build'
  GEN     ./Makefile
scripts/kconfig/conf  --silentoldconfig Kconfig
  Using .. as source for kernel
  GEN     ./Makefile
  CC      src/init/main.o
  CC      src/utils/src/cfassert.o
  CC      src/hal/src/freeRTOSdebug.o
  CC      src/hal/src/ledseq.o
  CC      src/hal/src/pm_stm32f4.o
python3 ../tools/make/versionTemplate.py --crazyflie-base .. ../src/utils/src/version.vtpl src/utils/src/version_gen.c
  CC      src/hal/src/radiolink.o
  CC      src/hal/src/sensors_bmi088_bmp3xx.o
  CC      src/drivers/src/led.o
  LD      src/init/built-in.o
  CC      src/deck/drivers/src/usddeck.o
  CC      src/hal/src/sensors_mpu9250_lps25h.o
  CC      src/drivers/src/nvic.o
  CC      src/modules/src/serial_4way.o
  CC      src/hal/src/usblink.o
  CC      src/modules/src/system.o
  CC      src/modules/src/cpx/cpxlink.o
  LD      src/drivers/src/built-in.o
  LD      src/deck/drivers/src/built-in.o
  LD      src/hal/src/built-in.o
  LD      src/drivers/built-in.o
  LD      src/hal/built-in.o
  LD      src/deck/built-in.o
  LD      src/modules/src/cpx/built-in.o
  LD      src/modules/src/built-in.o
  LD      src/modules/built-in.o
  CC      src/utils/src/version_gen.o
  LD      src/utils/src/version.o
  LD      src/utils/src/built-in.o
  LD      src/utils/built-in.o
  LD      src/built-in.o
  LD      firmware
  COPY  cf2.bin
  COPY  cf2.hex
Build for the cf2 platform!
Build 161:ec6a7929d5b1 (2023.11 +161) **MODIFIED**
Version extracted from git
Flash |  279948/1032192 (27%),  752244 free | text: 272544, data: 7404, ccmdata: 0
RAM   |   82336/131072  (63%),   48736 free | bss: 74932, data: 7404
CCM   |   62040/65536   (95%),    3496 free | ccmbss: 62040, ccmdata: 0
make[1]: Leaving directory '/mnt/c/Users/NAME/crazyflie-firmware/build'
Version extracted from git
Flash |  279948/1032192 (27%),  752244 free | text: 272544, data: 7404, ccmdata: 0
RAM   |   82336/131072  (63%),   48736 free | bss: 74932, data: 7404
CCM   |   62040/65536   (95%),    3496 free | ccmbss: 62040, ccmdata: 0
make[1]: Leaving directory '/mnt/c/Users/NAME/crazyflie-firmware/build'

I'm worried about the MODIFIED note here, but let me know if this is how it is supposed to be. 

When I run the command to flash, this is what I get:

root@NAME:/mnt/c/Users/NAME/crazyflie-firmware# CLOAD_CMDS="-w radio://0/80/2M" make cload
tools/kbuild/Makefile.kbuild:147: warning: overriding recipe for target 'cload'
Makefile:140: warning: ignoring old recipe for target 'cload'
make[1]: Entering directory '/mnt/c/Users/NAME/crazyflie-firmware/build'
python.exe -m cfloader -w radio://0/80/2M flash  cf2.bin stm32-fw
Reset to bootloader mode ...
nRF51 has: sd-s130 and requires None and upgrade provides None. Current bootloader version is [2024.2.0] but upgrade provides [None]
Flashing 1 of 1 to stm32 (fw): 280403 bytes (274 pages) ..........10..........10..........10..........10..........10..........10..........10..........10..........10..........10..........10..........10..........10..........10..........10..........10..........10..........10..........10..........10..........10..........10..........10..........10..........10..........10..........10....4

make[1]: Leaving directory '/mnt/c/Users/NAME/crazyflie-firmware/build'

However, when I turn the drone on, the code I changed doesn't work. Rather than the M1 light changing from red to green, there is no light at all.

**Reply by evoggy** [reply]
If you revert your changes and build/flash again does it work?

**Reply by cuber-1** [reply]
It does work again if I build/flash again. The red light blinks at a continuous rate.

**Reply by evoggy** [reply]
So if I understand correctly if you make the change and flash it no LEDs go on (not even M2) but if you revert the changes everything works? If this is the case it a bit strange since M2 is controlled by the nRF so it should not be effected. Can you still scan/connect to the Crazyflie from the client if you flash the example changes?

**Reply by cuber-1** [reply]
Yes, it flies perfect. Just to make sure this wasn't a drone issue, I went ahead and installed the VM to see if anything different happens. The modified code works on the VM - the light flashes green instead of red. I'll describe a bit more of what I saw on my Windows (WSL). both the M1 and M4 lights just didn't show at all.

**Comment by cuber-1** [comment]
Update: I unplugged the drone battery, plugged it back, turned the drone on, turned the drone off, held the power button to enter bootloader mode, and everything seems to work fine now. Thanks for all your help guys!

**Reply by evoggy** [reply]
Good to hear!
